### PR TITLE
Ability to disable auto-create

### PR DESCRIPTION
**Pull request by davidzhao**
WIP

### CODE REVIEW

File: pkg/service/roomservice.go [hunk 1]
```diff
@@ -75,6 +81,21 @@ func (s *RoomService) DeleteRoom(ctx context.Context, req *livekit.DeleteRoomReq
 		return nil, err
 	}
 
+	// we should not return until when the room is confirmed deleted
+	err = confirmExecution(func() error {
```
Comment (davidzhao) [inline pkg/service/roomservice.go:85]: room service functions had a shortcoming that they returned too soon before the operation has actually completed on the media node. this has led to unpredictable behavior for users but also our tests.

File: test/singlenode_test.go [hunk 2]
```diff
@@ -325,3 +327,57 @@ func TestSingleNodeCORS(t *testing.T) {
...
+		defer s.Stop(true)
+
+		waitForServerToStart(s)
+
+		token := joinToken(testRoom, "start-before-create")
+		_, err := testclient.NewWebSocketConn(fmt.Sprintf("ws://localhost:%d", defaultServerPort), token, nil)
+		require.Error(t, err)
+	})
+
+	t.Run("join with explicit creatRoom", func(t *testing.T) {
```
Comment (boks1971) [inline test/singlenode_test.go:363]: :nit: `createRoom`

File: test/scenarios.go [hunk 3]
```diff
@@ -141,6 +140,22 @@ func scenarioDataPublish(t *testing.T) {
...
+	// close room with room client
+	_, err := roomClient.DeleteRoom(contextWithToken(createRoomToken()), &livekit.DeleteRoomRequest{
+		Room: testRoom,
+	})
+	require.NoError(t, err)
+
+	// now join again
+	c2 := createRTCClient("jcr2", defaultServerPort, nil)
+	waitUntilConnected(t, c2)
+	defer stopClients(c2)
```
Comment (boks1971) [inline test/scenarios.go:156]: why only `c2`? Also, why `defer`?

File: test/singlenode_test.go [hunk 4]
```diff
@@ -325,3 +327,57 @@ func TestSingleNodeCORS(t *testing.T) {
...
+				logger.Errorw("server returned error", err)
+			}
+		}()
+		defer s.Stop(true)
+
+		waitForServerToStart(s)
+
+		token := joinToken(testRoom, "start-before-create")
+		_, err := testclient.NewWebSocketConn(fmt.Sprintf("ws://localhost:%d", defaultServerPort), token, nil)
+		require.Error(t, err)
```
Comment (boks1971) [inline test/singlenode_test.go:360]: Don't understand what this is testing.

Should this actually create a `testRTCClient` and check that it gets a 404?

**Review by boks1971 — APPROVED** [review]
A couple of questions and some minor comments.

Can't wait to have more stable tests :-)

File: pkg/config/config_test.go [hunk 5]
```diff
@@ -13,3 +13,12 @@ func TestConfig_UnmarshalKeys(t *testing.T) {
 	require.NoError(t, conf.unmarshalKeys("key1: secret1"))
 	require.Equal(t, "secret1", conf.Keys["key1"])
 }
+
+func TestConfig_DefaultsKept(t *testing.T) {
+	const content = `room:
+  empty_timeout: 10`
```
[the remark this answers is not in the document]
Comment (davidzhao) [reply pkg/config/config_test.go:19]: this is intentional, to be YAML:

```
room:
  empty_timeout: 10
```

File: pkg/service/rtcservice.go [hunk 6]
```diff
@@ -23,18 +24,23 @@ import (
...
+	store         RoomStore
 	upgrader      websocket.Upgrader
 	currentNode   routing.LocalNode
+	config        *config.Config
 	isDev         bool
 	limits        config.LimitConfig
 }
 
-func NewRTCService(conf *config.Config, ra RoomAllocator, router routing.MessageRouter, currentNode routing.LocalNode) *RTCService {
+func NewRTCService(conf *config.Config, ra RoomAllocator, store RoomStore,
```
[the remark this answers is not in the document]
Comment (davidzhao) [reply pkg/service/rtcservice.go:35]: sounds good, I like that as a style guideline

File: test/singlenode_test.go (hunk 4 above)
Comment (davidzhao) [reply test/singlenode_test.go:360]: it's testing that it does not auto-create the room (and thus fail the join websocket connection)

currently the test client will panic when it fails. The websocket connection logic is the first bit of what createRTCClient is doing.

**Pull request merged**